tinygrad/runtime/ops_hip.py

```diff
@@ -63,22 +77,55 @@ def __init__(self, name:str, prg:bytes):
-    if wait:
-      hip.hipEventRecord(end)
-      hip.hipEventSynchronize(end)
-      ret = hip.hipEventElapsedTime(start, end)*1e-3
-      hip.hipEventDestroy(start)
-      hip.hipEventDestroy(end)
-      return ret
+    return time_execution(lambda: hip.hipModuleLaunchKernel(self.prgs[args[0]._device], global_size[0], global_size[1], global_size[2], local_size[0], local_size[1], local_size[2], 0, 0, struct), enable=wait)
 
   def __del__(self):
     for module in self.modules: hip.hipModuleUnload(module)
 
+class HIPGraph(BatchExecutor):
+  def __init__(self, jit_cache: List[JitItem], input_rawbuffers: List[RawBuffer], var_vals: Dict[Variable, int]):
+    super().__init__(jit_cache, input_rawbuffers, var_vals)
+
+    self.graph, graph_node = hip.hipGraphCreate(), None
+    self.updatable_nodes: Dict[int, Tuple[Any, hip.kernelNodeParamsWrapper]] = {} # Dict[jc index] = tuple(graph_node, node_params)
+
+    for (j,i),input_name in self.input_replace.items(): self.jit_cache[j].rawbufs[i] = input_rawbuffers[input_name]
+    for j,ji in enumerate(self.jit_cache):
+      prg: CompiledASTRunner = cast(CompiledASTRunner, ji.prg)
+      global_size, local_size = prg.launch_dims(var_vals)
+
+      assert all(x is not None for x in ji.rawbufs) and ji.rawbufs[0] is not None, "buffers could not be None" # for linters
+      params = hip.buildKernelNodeParams(*ji.rawbufs, *[var_vals[x] for x in prg.vars], func=prg.clprg.prgs[ji.rawbufs[0]._device], grid=global_size, block=local_size)
+      graph_node = hip.hipGraphAddKernelNode(self.graph, [graph_node] if graph_node else [], params)
+
+      if j in self.jc_idxs_with_updatable_launch_dims or j in self.jc_idxs_with_updatable_var_vals or j in self.jc_idxs_with_updatable_rawbufs:
+        self.updatable_nodes[j] = (graph_node, params)
+
+    self.instance = hip.hipGraphInstantiate(self.graph)
+    self.clear_jit_inputs()
+
+  def __del__(self):
+    if hasattr(self, 'instance'): hip.hipGraphExecDestroy(self.instance)
+    if hasattr(self, 'graph'): hip.hipGraphDestroy(self.graph)
+
+  def __call__(self, input_rawbuffers: List[RawBuffer], var_vals: Dict[Variable, int], wait=False):
+    # Update cached params structs with the new values.
+    for (j,i),input_idx in self.input_replace.items():
+      hip.setKernelNodeParam(self.updatable_nodes[j][1], input_rawbuffers[input_idx], i)
+    for j in self.jc_idxs_with_updatable_launch_dims:
+      hip.setKernelNodeLaunchDims(self.updatable_nodes[j][1], cast(CompiledASTRunner, self.jit_cache[j].prg).launch_dims(var_vals))
+    for j in self.jc_idxs_with_updatable_var_vals:
+      hip.setKernelNodeParams(self.updatable_nodes[j][1], [var_vals[x] for x in self.jit_cache[j].prg.vars], list(range(len(self.jit_cache[j].rawbufs), len(self.jit_cache[j].rawbufs) + len(self.jit_cache[j].prg.vars))))
+
+    # Update graph nodes with the updated structs.
+    for j in self.updatable_nodes:
```

Comment:
`for node, params: hip.hipGraphExecKernelNodeSetParams(self.instance, node, params)`